src/procedures/proc_spmw_paths.c

```diff
@@ -0,0 +1,471 @@
+				SIValue w = _get_value_or_defualt((GraphEntity *)&frontierConnection.edge, ctx->weight_prop, SI_LongVal(1));
+				if(p->cost + SI_GET_NUMERIC(c) <= ctx->max_cost && p->weight + SI_GET_NUMERIC(w) <= max_weight) {
+					p->cost += SI_GET_NUMERIC(c);
+					p->weight += SI_GET_NUMERIC(w);
+					Path_AppendEdge(allPathsCtx->path, frontierConnection.edge);
+				} else {
+					Path_PopNode(allPathsCtx->path);
+					continue;
+				}
+			}
+
+			// Update path depth.
+			depth++;
+
+			/* Introduce neighbors only if path depth < maximum path length.
+			 * and frontier wasn't already expanded. */
+			if(depth < allPathsCtx->maxLen && !frontierAlreadyOnPath) {
+				addNeighbors(allPathsCtx, &frontierConnection, depth, allPathsCtx->dir);
+			}
+
+			// See if we can return path.
+			if(depth >= allPathsCtx->minLen && depth <= allPathsCtx->maxLen) {
+				if(allPathsCtx->dst != NULL) {
+					Node dst = Path_Head(allPathsCtx->path);
+					if(ENTITY_GET_ID(allPathsCtx->dst) != ENTITY_GET_ID(&dst)) continue;
+				}
+				p->path = allPathsCtx->path;
+				return;
+			}
+		} else {
+			// No way to advance, backtrack.
+			Path_PopNode(allPathsCtx->path);
+			if(Path_EdgeCount(allPathsCtx->path)) {
+				Edge e = Path_PopEdge(allPathsCtx->path);
+				SIValue c = _get_value_or_defualt((GraphEntity *)&e, ctx->cost_prop, SI_LongVal(1));
+				SIValue w = _get_value_or_defualt((GraphEntity *)&e, ctx->weight_prop, SI_LongVal(1));
+				p->cost -= SI_GET_NUMERIC(c);
+				p->weight -= SI_GET_NUMERIC(w);
+			}
+		}
+	}
+
+	// Couldn't find a path.
+	p->path = NULL;
+	return;
+}
+
+static int path_cmp(const void *a, const void *b, const void *udata) {
+	WeightedPath *da = (WeightedPath *)a;
+	WeightedPath *db = (WeightedPath *)b;
+	if(da->weight == db->weight) {
+		if(da->cost - db->cost) {
+			return Path_Len(da->path) - Path_Len(db->path);
+		}
+		return da->cost - db->cost;
+	}
+	return da->weight - db->weight;
+}
+
+static void SPMWpaths_all_minimal
+(
+	SPMWctx *spmw_ctx
+) {
+	spmw_ctx->array = array_new(WeightedPath, 0);
+
+	WeightedPath p = {0};
+	double max_weight = DBL_MAX;
+	SPMWpaths_next(spmw_ctx, &p, max_weight);
+	while (p.path != NULL) {
+		uint count = array_len(spmw_ctx->array);
+		if(count > 0 && p.weight < spmw_ctx->array[0].weight) {
+			for (uint i = 0; i < array_len(spmw_ctx->array); i++) {
+				Path_Free(spmw_ctx->array[i].path);
+			}
+			array_clear(spmw_ctx->array);
+			count = 0;
+		}
+		p.path = Path_Clone(p.path);
+		array_append(spmw_ctx->array, p);
+		max_weight = p.weight;
+		SPMWpaths_next(spmw_ctx, &p, max_weight);
+	}
+}
+
+static void SPMWpaths_single_minimal
+(
+	SPMWctx *spmw_ctx
+) {
+	spmw_ctx->single.path   = NULL;
+	spmw_ctx->single.weight = DBL_MAX;
+	spmw_ctx->single.cost   = DBL_MAX;
+	WeightedPath p = {0};
+	SPMWpaths_next(spmw_ctx, &p, DBL_MAX);
+	while (p.path != NULL) {
+		if(p.weight < spmw_ctx->single.weight ||
+			(p.weight == spmw_ctx->single.weight &&
```

Comment:
due to `SPMWpaths_next` optimization, found path is guarantee to be <= current path. as such `p.weight == spmw_ctx->single.weight` can be omitted